### ISSUE TITLE
[jvm-packages] xgboost4j-spark: all predictions 0

### ISSUE THREAD

**Issue by YCG09** [dmlc/xgboost#2297]
Operating System: Centos7.0  (with Spark 2.1.1)
Package: jvm, xgboost4j-spark
`xgboost` version used: 0.7 (May 9th)

Recently, I use xgboost4j-spark for text multi-classification on Pipline and crossValidate, and I got all predictions 0 no mastter I modified the parameters.

Today, I find a similar problem #1941 , and I changed the example code for multi-classification, as below:

import org.apache.spark.ml.Pipeline
import ml.dmlc.xgboost4j.scala.spark.XGBoostEstimator
import org.apache.spark.ml.evaluation.BinaryClassificationEvaluator
import org.apache.spark.ml.feature.{HashingTF, Tokenizer}
import org.apache.spark.ml.linalg.Vector
import org.apache.spark.ml.tuning.{CrossValidator, ParamGridBuilder}
import org.apache.spark.sql.Row
import scala.collection.mutable

// xgboost parameters
def get_param(): mutable.HashMap[String, Any] = {
  val params = new mutable.HashMap[String, Any]()
  params += "eta" -> 0.1
  params += "max_depth" -> 8
  params += "gamma" -> 0.0
  params += "colsample_bylevel" -> 1
  params += "objective" -> "multi:softmax"
  //params += "objective" -> "multi:softprob"
  params += "num_class" -> 3
  params += "booster" -> "gbtree"
  params += "num_rounds" -> 20
  return params
}

// Prepare training data from a list of (id, text, label) tuples.
val training = spark.createDataFrame(Seq(
  (0L, "a b c d e spark", 1.0),
  (1L, "b d", 0.0),
  (2L, "spark f g h", 1.0),
  (3L, "hadoop mapreduce", 0.0),
  (4L, "b spark who", 1.0),
  (5L, "g d a y", 0.0),
  (6L, "spark fly", 1.0),
  (7L, "was mapreduce", 0.0),
  (8L, "e spark program", 1.0),
  (9L, "a e c l", 0.0),
  (10L, "spark compile", 1.0),
  (11L, "hadoop software", 0.0),
  (12L, "hive task", 2.0),
  (13L, "use hive", 2.0),
  (14L, "good luck hive", 2.0),
  (15L, "hello hive", 2.0)
)).toDF("id", "text", "label")

// Prepare test documents, which are unlabeled (id, text) tuples.
val test = spark.createDataFrame(Seq(
  (4L, "spark i j k"),
  (5L, "l m n"),
  (6L, "mapreduce spark"),
  (7L, "apache hadoop"),
  (8L, "hive test")
)).toDF("id", "text")

// Configure an ML pipeline, which consists of three stages: tokenizer, hashingTF, and xgb.
val tokenizer = new Tokenizer().setInputCol("text").setOutputCol("words")
val hashingTF = new HashingTF().setInputCol(tokenizer.getOutputCol).setOutputCol("features")

val xgb = new XGBoostEstimator(get_param().toMap).setLabelCol("label").setFeaturesCol("features").setPredictionCol("prediction")

val pipeline = new Pipeline() .setStages(Array(tokenizer, hashingTF, xgb))

// test pipeline
val xgbModel = pipeline.fit(training)
xgbModel.transform(test).select("text", "probabilities", "prediction").show()

And finally I got this:
+---------------+-------------+----------+                                      
|                  text| probabilities|prediction|
+---------------+-------------+----------+
|    spark i j k     |              [1.0]|           0.0|
|          l m n      |              [0.0]|           0.0|
|mapreduce spark|          [1.0]|           0.0|
|  apache hadoop|           [0.0]|           0.0|
|      hive test     |              [2.0]|           0.0|
+---------------+-------------+----------+

It looks like the prediction col is probabilities, and when I changed 'objective' to "multi:softprob", it looks ok
+---------------+--------------------+----------+                               
|                  text|            probabilities|prediction|
+---------------+--------------------+----------+
|        spark i j k|[0.31553599238395...|          1.0|
|               l m n|[0.37241610884666...|          0.0|
|mapreduce spark|[0.31553599238395...|     1.0|
|  apache hadoop|[0.37241610884666...|      0.0|
|          hive test|[0.31829428672790...|         2.0|
+---------------+--------------------+----------+

So, I think the prediction col for "multi:softmax" has some problem, and setting for this is useless, it‘s also the reason I got all prediction 0.

**Comment by CodingCat** [comment]
would you please add [jvm-packages] in the title when you ask about jvm-packages related questions?

**Comment by CodingCat** [comment]
did you try multi:softmax with other bindings?

**Comment by CodingCat** [comment]
"probabilities" is the default col name of the prediction, when you set multi:softmax as objective, you directly get prediction instead of probabilities as output

when you predict over the prediction result (taking it as probabilities)...you will get some unexpected results

**Comment by YCG09** [comment]
Ok，added.
I got it, but why setPredictionCol method is useless?

**Comment by CodingCat** [comment]
no, proabilities is the default name of the `rawprediction`, i.e. the probabilities which determine you say it is 0, 1, 2, 3, 4....

The workflow is showing in here: https://github.com/dmlc/xgboost/blob/master/jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostClassificationModel.scala#L149

when you set multi:softmax, that means you do not need rawprediction appear in your DF, then you can set rawPrediction as empty to get "prediction" column directly

**Comment by YCG09** [comment]
I get it, thanks for your patience.
